### ISSUE TITLE
State does not change properly during testing

### ISSUE THREAD

**Issue by convers39** [viewflow/django-fsm#249]
Hi there,
I am working on a project and use this package to manage the order status.
I added protect=True in my model:
```python
status = FSMField(_("order status"), choices=Status.choices, default=Status.NEW, protected=True)
```
It works well, however, when I run a test, the status does not change as expected all the time:

**Case A** - run FSM method **directly** in the test, status changed as expected, the test will pass
```python
# models.py
    @transition(field=status, source=['CL'], target='CX')
    def confirm_cancel(self):
        return True

# test_models.py
    def test_cancel_restore_product_stock(self):
        order = OrderFactory(status='CL')
        sku = self.skus[0]
        op = OrderProductFactory(product=sku)
        op.order = order
        op.save()
        order.confirm_cancel()
        order.restore_product_stock()
        sku.refresh_from_db()

        **self.assertEqual(order.status, 'CX')**
        self.assertEqual(sku.stock, 6)
        self.assertEqual(sku.sales, 0)

```

**Case B** - run fsm method **indirectly** (trigger in a view method), status did not change
```python
# models.py 
    @transition(field=status, source='SP', target='RT', conditions=[in_return_deadline])
    def request_return(self):
        self.return_at = datetime.now()

# views.py
    def post(self, request, *args, **kwargs):
        data = json.loads(request.body.decode())
        order = Order.objects.get(id=data['order_id'])

        status = order.status
        if status == 'SP':
            order.request_return()
            order.save()
            return JsonResponse({'res': '1', 'msg': 'Return request has been sent'})
    # ...

# test_views.py
    def test_request_return(self):
        order = OrderFactory(user=self.user, status='SP')
        self.client.force_login(self.user)
        payload = {'order_id': order.id}
        res = self.client.post(self.url, data=payload,
                               content_type=self.content_type)
        self.assertEqual(res.status_code, 200)
        self.assertEqual(
            res.json(), {'res': '1', 'msg': 'Return request has been sent'})
        self.assertEqual(order.status, 'RT')
```
This will lead to a failure:
```bash
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/src/app/apps/order/tests/test_views.py", line 218, in test_request_return
    self.assertEqual(order.status, 'RT')
AssertionError: 'SP' != 'RT'
- SP
+ RT
```

Now there are two different ways to make the status change as expected:

**a**. If I remove the `protect=True` kwarg in my model, then run `order.refresh_from_db()` in the test, the test will pass.

**b**. If I run the `request_return` method in Case B test directly, the test will pass, but this does not make sense if I have to manually run the FSM method in every test.

Does anyone have any idea about this issue? 
Or is there any way to **override `protect=True`** then i may use `refresh_from_db` in the test ?

**Comment by kmmbvnr** [comment]
refresh_from_db with protected=True are not supported, see #174 

I don't think it could be implemented in any reasonable way, without several complicated hacks.

Since, I'm going soon to deprecate django-fsm, I would rather suggest to switch to my new viewflow.fsm package - https://docs-next.viewflow.io/fsm/index.html (already available in 2.0.0a0 release)

`viewflow.fsm` separates django model and flows, and solves a lot of issues, including refresh_from_db, deferred fields and transition method chaining

**Comment by convers39** [comment]
Hi @kmmbvnr 
Thank you for your immediate response.
Noted with thanks, I will transfer to viewflow then.

As a temporary solution I just assign protect dynamically..
```python
status = FSMField(_("order status"),
                      choices=Status.choices, default=Status.NEW,
                      protected=False if os.environ['DJANGO_SETTINGS_MODULE'].split('.')[-1] == 'testing' else True)
```
